persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -360,20 +364,31 @@ openSimpleConn = openSimpleConnWithVersion getServerVersion
 -- @since 2.9.1
 openSimpleConnWithVersion :: (PG.Connection -> IO (Maybe Double)) -> LogFunc -> PG.Connection -> IO SqlBackend
 openSimpleConnWithVersion getVerDouble logFunc conn = do
-    smap <- newIORef $ Map.empty
+    smap <- mkSimpleStatementCache
     serverVersion <- oldGetVersionToNew getVerDouble conn
-    return $ createBackend logFunc serverVersion smap conn
+    return $ createBackend logFunc serverVersion (mkStatementCache smap) conn
+
+underlyingConnectionKey :: Vault.Key PG.Connection
+underlyingConnectionKey = unsafePerformIO Vault.newKey
+{-# NOINLINE underlyingConnectionKey #-}
```

Comment:
This doesn't feel right to me. My understanding of this behavior is:

1. We create this `Vault.Key PG.Connection` lazily, whenever the first time we call this function. Then the value is cached due to GHC's applicative-normal-form optimization. So, we generate exactly one `Key Connection` here, and it is a global constant.
2. When we call `createBackend` or `getSimpleConn` (whichever one comes first) then we create the `underlyingConnectionKey`. If we call `getSimpleConn` first, then we'll get `Nothing` in that function. If we call `createBackend` afterwards, then the `Key Conn` already exists, and we stuff the `Conn` into the `Vault`.
3. So, I think the problem with this, is that we are *overwriting* the underlying `Conn` with every call to `createBackend`. Is this a problem?

`withPostgresqlPool` delegates to `withSqlPool open'` which itself calls `return $ constructor (createBackend logFunc ver smap) conn`. Which means that, on every new `SqlBackend` that is created, we end up overwriting the value here, which means that `getSimpleConn` is going to get the most *recent* `PG.Conn` that has been created.

Reply:
OK, so this is probably not an issue. While we do generate a single `Key Conn`, this is used to index into the `SqlBackend`'s internal `Vault`, not some globally available `Vault`. And, since there's only ever exactly one, then we can easily recover the `PG.Connection` for a given `SqlBackend`.

This is a bit gnarly but it shouldn't have any problems.

Reply:
Points 1 and 2 are intentional. We need a stable key in order to be able to reference the connection from external instrumentation tooling. AFAICT, there's no way you could call `getSimpleConn` prior to constructing a backend via `createBackend`, so you're guaranteed by virtue of the order of operations to have the key be initialized and able to return the underlying PostgreSQL connection. 

Regarding 3, I don't see how we're overwriting the underlying call in `createBackend`, since it's just an immutable function that constructs a backend from its constituent parts?